[MRG+1] Per-key priorities for dict-like settings by promoting dicts to Settings instances

jdemaeyer opened the pull request:
Expand settings priorities by assigning per-key priorities for the dict-like settings (e.g. `DOWNLOADER_MIDDLEWARES`), instead of just a single priority for the whole dictionary. This allows updating these settings from multiple locations without having to care (too much) about order. It is a prerequisite for the add-on system (#1272, #591).

There are two main updates:
 1. All default dictionary settings are promoted to the `BaseSettings` class (formerly `Settings`). They behave just like dictionaries, but honour per-key priorities when being written to.
 2. Since their rationale becomes obsolete with per-key priorities, all `X_BASE` settings are deprecated, with default entries now living in the `X` setting.

And several smaller updates:
 1. `Settings` is a subclass of `BaseSettings`. It loads the default settings and promotes dictionaries within them to `BaseSettings` instances.
 2. `BaseSettings` has a complete dictionary-like interface.
 3. *All* dictionary-like component/handler lists allow disabling components/handler by setting the value of a key/value pair to `None`. A new helper, `scrapy.util.without_none_values()` was introduced for this. This was previously not supported by `FEED_STORAGES`, `FEED_EXPORTERS`, and `DEFAULT_REQUEST_HEADERS`.
 4. The `scrapy.util.build_component_list()` helper has been updated according to the deprecation of `_BASE` settings, as the `(base, custom)` call signature does not make much sense anymore. ~~It's still backwards-compatible.~~
 5. **`ITEM_PIPELINES` can no longer be provided as list**

Comes with many new/updated tests and documentation.

jdemaeyer commented:
Implementing per-key priorities like this would render #1110 obsolete

jdemaeyer commented:
I've resolved issue 1 by providing a `BaseSettings` class that contains all the methods. `Settings` is now a subclass which overwrites `__init__()` and loads the default setting.
Issue 3 is resolved by completely resolving a `Settings`-setting when the update value is not a mapping, and I've outsourced the duplicate priority string resolving (issue 4).
Issue 5 should probably get its own PR.
@Curita I needed to merge master into this since automatic merging was no longer possible at some point. But now I have this huge commit in this PR. Do I wait until we're ready to merge this and then rebase before we do?

Curita commented on scrapy/core/downloader/handlers/__init__.py:
```diff
@@ -13,7 +13,7 @@ class DownloadHandlers(object):
         self._handlers = {}
         self._notconfigured = {}
-        handlers = crawler.settings.get('DOWNLOAD_HANDLERS_BASE')
+        handlers = crawler.settings.get('DOWNLOAD_HANDLERS_BASE', {})
```
We could use `crawler.settings.getdict()` (which has `{}` as default already) to get both 'DOWNLOAD_HANDLERS_BASE' and 'DOWNLOAD_HANDLERS'.

nramirezuy replied:
Not just that it also allows you to pass `DOWNLOAD_HANDLERS` from command line.

Curita commented on scrapy/extensions/feedexport.py:
```diff
@@ -195,8 +195,7 @@ def item_scraped(self, item, spider):
     def _load_components(self, setting_prefix):
-        conf = dict(self.settings['%s_BASE' % setting_prefix])
-        conf.update(self.settings[setting_prefix])
+        conf = dict(self.settings[setting_prefix])
```
Shouldn't we keep '_BASE' loading an updating here for backward compatibility as the other cases?

Curita replied:
Also it's probably better to replace that call for `conf = self.settings.getdict(settings_prefix)`.

Curita commented on scrapy/settings/__init__.py:
```diff
@@ -19,6 +19,12 @@
     'cmdline': 40,
 }
 
+def get_settings_priority(priority):
```
:+1:

Curita commented on scrapy/settings/__init__.py:
```diff
@@ -34,9 +40,20 @@ def __init__(self, value, priority):
+            #      that does not map keys to values, e.g.a list). I included
+            #      this b/c the test_cmdline/__init__.py test uses a deprecated
+            #      configuration API (EXTENSIONS is a list) and fails otherwise.
+            #      Should it stay in?
```
Let's remove `and isinstance(value, Mapping)`, `build_component_list` already handles the case of `custom` being a list or a tuple by returning it right away, without updating '_BASE'. Other settings that don't use `build_component_list` should break since passing lists wasn't an option before.

jdemaeyer replied:
The thing is that the `EXTENSIONS` test was only an example of what developers might actually *want* to do. What if they have a settings that is a `BaseSettings` instance, but for some reason they decide to replace it with a different kind of variable, say `None`. The `isinstance(value, Mapping)` allows for this.
I guess the alternative is to make developers actively acknowledge that they're losing the `BaseSettings` instance and all its values/priorites by deleting it, then inserting a new setting with the type they want. While that is more explicit, it's probably not expected behaviour, given that "type replacement" for all other types of settings does not yield exceptions.

jdemaeyer replied:
On second thought, now that only the default dicts are promoted, and given that these should never be replaced by any incompatible type, I will remove the `and isinstance(value, Mapping`). It also causes problems with passing json-encoded strings.

Curita commented on scrapy/settings/__init__.py:
```diff
@@ -34,9 +40,20 @@ def __init__(self, value, priority):
+            #      configuration API (EXTENSIONS is a list) and fails otherwise.
+            #      Should it stay in?
+            self.value.update(value, priority)
+            self.priority = max((priority, self.priority))
```
Idea here is that `self.priority` is the highest priority within the internal `BaseSettings`? In that case, I think we should lookup inside the provided `value` instead of relying in the passed `priority`. Examples like this one won't get the expected priority otherwise:

```python
> s = Settings()
> s.set('option', BaseSettings({'key': 'value'}, priority='spider'))
> s.attributes['option'].priority
  20  # priority 'project' instead of 'spider'
```

jdemaeyer replied:
Yep, that was the idea. I've updated the code so it works as expected, thereby introducing a `maxpriority()` method to the `BaseSettings` class since there are two places in the code where I need to find the maximum priority. It's probably not very useful for anything else though, does it clutter up the API too much?

Curita commented on scrapy/settings/__init__.py:
```diff
@@ -45,14 +62,12 @@ def __str__(self):
 
 
-class Settings(object):
+class BaseSettings(MutableMapping):
```
I really like this change :+1:

Curita commented on scrapy/settings/__init__.py:
```diff
@@ -88,19 +106,30 @@ def getdict(self, name, default=None):
             value = json.loads(value)
         return dict(value)
 
+    def getpriority(self, name):
```
Needed method :+1:

Curita commented on scrapy/settings/__init__.py:
```diff
@@ -88,19 +106,30 @@ def getdict(self, name, default=None):
+            else:
+                if isinstance(value, dict):
+                    value = BaseSettings(value, priority)
+                self.attributes[name] = SettingsAttribute(value, priority)
```
After giving it some more thought, I think we should deal with BaseSettings instances explicitly instead of promoting dictionaries. That way users won't get unexpected or backward incompatible behaviors in their own dict settings, and we can choose specifically which settings are going to be BaseSettings by just declaring their values to be that.

We can still do some sort of implicit propagation only in the `Settings` class while loading the default values, checking if a setting is a dict and replacing or setting it (depending when it's done) to a BaseSettings object instead. Another option if that's not possible could it be to just import BaseSettings into `default_settings.py` and use it there for the dict values. For that matter maybe we could move `Settings` to another module within `scrapy/settings/` so we get rid of the default_settings import in scrapy/settings/\__init__.py.

jdemaeyer replied:
I agree. Now that we split `Settings` and `BaseSettings` explicit promotion of only the default settings seems a more sane choice.

Curita commented on scrapy/settings/__init__.py:
```diff
@@ -110,6 +139,25 @@ def setmodule(self, module, priority='project'):
+            del self.attributes[name]
+
+    def __delitem__(self, name):
+        self.delete(name)
```
I'm not sure I like the new `delete()` method, seems like users could get confused about their usage and their differences and similarities with `set()`. For keeping the dict-like interface I think we could keep `__delitem__` with a simpler implementation, such as deleting the mapping in `self.attributes`.

nramirezuy replied:
Shouldn't it also consider `frozen` ?

jdemaeyer replied:
I like having the option of "Delete this setting *unless someone has updated it with a higher priority than what I'm doing*", but would yield to the majority if that is considered useless. ;)
But I think you're right in that `__delitem__` should ignore priorities and just delete from `attributes`

Curita commented on scrapy/utils/conf.py:
```diff
@@ -13,7 +15,7 @@ def build_component_list(base, custom):
     if isinstance(custom, (list, tuple)):
         return custom
-    compdict = base.copy()
+    compdict = BaseSettings(base, priority = 'default')
```
A note about coding style in this line and in general: though we're not enforcing it we try to comply with [pep8](https://www.python.org/dev/peps/pep-0008/) for new added code to scrapy as closely as we can (only rule we agreed to change it's the 80 chars max line width rule, we settled on a more flexible 100 chars). This pull request follows pep8 mostly, but I'd try to stick to the rule of "There shouldn't be spaces around default arguments nor immediately after parenthesis, brackets or braces".

jdemaeyer replied:
I definitely try to stick to PEP8, but I have to admit that I frequently break the "no spaces for keyword arguments" rule b/c it feels so unnatural to me. Will fix :)

Curita commented:
> Currently DOWNLOAD_HANDLERS is the only setting that does not use the dictionary values for ordering (but instead uses them to store the path of handler classes) and thus cannot use build_component_list(). However, this may change in the future, maybe most of the code from DownloadHandlers.\__init__() should be outsourced into a new helper function (similar to build_component_list but without ordering) in scrapy.utils.conf?

FEED_STORAGES and FEED_EXPORTERS are dictionaries with paths that don't use ordering too, and those should handle \*\_BASE settings (DEFAULT_REQUEST_HEADERS is a dict as well, but it's used differently). I like the idea of outsourcing the common code from FEED_* and DOWNLOAD_HANDLERS loading and `build_component_list`, and it makes sense to do so in this pr. Particularly I'd like to see a single place handling the \<prefix>\_BASE setting loading and updating with \<prefix> setting so we can deprecate it sometime in the future more easily. Another thing that's useful for is to ensure that all settings can be disabled by assigning them the value `None`. Right now I think both FEED_* settings don't support this for example (and they should).

> @Curita I needed to merge master into this since automatic merging was no longer possible at some point. But now I have this huge commit in this PR. Do I wait until we're ready to merge this and then rebase before we do?

Don't worry, just rebase scrapy's master when you can (we'll definitely need that before merging but can be done at any time you want) and force push the new git history.

jdemaeyer commented:
Thank you guys for the feedback!

I now have:
  1. Disabled auto-promotion of dicts, only dicts from `default_settings` will now be promoted to `BaseSettings` instances
  2. Updated the `SettingsAttribute.priority` behaviour for SettingsAttributes that contain a `BaseSettings` instance. It now always reflects the maximum of:
    - the highest priority in the `BaseSettings` instance
    - the highest priority ever given to `SettingsAttribute.set()` (even when that priority is higher than anything in `BaseSettings`)
  3. Updated `BaseSettings.update()` so it can also deal with JSON-encoded strings, this solves [the issue raised by](https://github.com/scrapy/scrapy/pull/1149#discussion_r31134086) @nramirezuy and avoids downgrading `BaseSettings` to `dicts` (losing all per-key priorities) wherever possible
  4. Deprecated `scrapy.utils.conf.build_component_list()` since the `(base, custom)` signature is no longer needed. I did not want to reuse the function name though since people might be using it in their extensions.
  5. As replacement, introduced `scrapy.utils.conf.build_components()` that only takes a *single* dict, removes all entries containing `None` as value, then builds the list or returns the dict, depending on the parameter `to_list=True`.
  6. Removed all references to `_BASE` settings (and their merging with the non-`_BASE` setting) in the complete codebase, instead replacing it with calls to  the function `scrapy.utils.conf._get_composite_setting(settings, settingname)`. This function, and the calls to it, should only be in the code until we decide to finally remove all support for the `_BASE` settings. I guess this is not exactly what @Curita had hoped for, since we will still have to replace all calls to `_get_composite_setting()` when we remove `_BASE` support, and replace them with a simple `settings[settingname]`. But the alternative (that I could see) would have been to include `settings` in the signature of `build_components()`, which seemed not very modular.
  7. Enabled disabling through setting a dict value to `None` for *all* default dict settings (including the default request headers) by placing calls to `build_components()` where appropriate.

jdemaeyer commented:
Alright, this PR is now at a stage where I'm fairly happy with it and would remove the WIP tag as soon as I've written/updated the documentation and rebased. Still very open for feedback of course :)

@Curita I made some changes to the `_map_keys()` function that lives inside of `build_component_list()` so that it honours per-key priorities. As giving both `base` and `custom` becomes deprecated with this PR though, and per-key priorities weren't available before anyways, maybe that's not necessary and just clutters up the code?

Curita commented on scrapy/settings/__init__.py:
```diff
@@ -174,11 +251,28 @@ def __iter__(self, k, v):
     def __init__(self, settings_module=None, **kw):
-        Settings.__init__(self, **kw)
         self.settings_module = settings_module
+        Settings.__init__(self, **kw)
```
There is a reason for swapping these two lines?

jdemaeyer replied:
Yes, took me quite a while to figure out though ;D It became necessary after I moved the dict promotion into `__init__()`. `Settings.__init__()` uses `__getitem__()` during the promotion of default dictionaries to `BaseSettings` instances, which in turn accesses `self.settings_module`, so it needs to be defined before calling `__init__()`

Curita commented on scrapy/settings/__init__.py:
```diff
@@ -88,19 +103,48 @@ def getdict(self, name, default=None):
+        # It's for internal use in the transition away from the _BASE settings and
+        # will be removed along with _BASE support in a future release
+        basename = name + "_BASE"
+        if basename in self.attributes:
```
nitpick: This line should be `if basename in self:` and other similar references could be replaced too.

Curita commented on scrapy/settings/__init__.py:
```diff
@@ -88,19 +103,48 @@ def getdict(self, name, default=None):
+            warnings.warn('_BASE settings are deprecated.',
+                          category=ScrapyDeprecationWarning)
+            compsett = BaseSettings(self[name + "_BASE"], priority='default')
+            compsett.update(self.get(name))
```
nitpick: `self[name]` instead of `self.get(name)`

jdemaeyer replied:
True. I forgot that `__getitem__()` doesn't throw `KeyError`s either. I like all your nitpicks ;)

Curita commented on scrapy/settings/__init__.py:
```diff
@@ -88,19 +103,48 @@ def getdict(self, name, default=None):
+
+    def maxpriority(self):
+        if len(self) > 0:
+            return max(self.getpriority(name) for name in self.attributes)
```
nitpick: `for name in self` instead of `for name in self.attributes`. I think there are other similar calls that could be replaced.

Curita commented on scrapy/settings/__init__.py:
```diff
@@ -88,19 +103,48 @@ def getdict(self, name, default=None):
+        if basename in self.attributes:
+            warnings.warn('_BASE settings are deprecated.',
+                          category=ScrapyDeprecationWarning)
+            compsett = BaseSettings(self[name + "_BASE"], priority='default')
```
I think getting this setting should be `self.getdict(name + "_BASE")` instead, the old deprecated _BASE could be a json dict.

jdemaeyer replied:
Since `__init__()` calls `update()`, it can deal with JSON strings

Curita commented on scrapy/utils/conf.py:
```diff
@@ -1,36 +1,54 @@
+            return type(custom)(convert(c) for c in custom)
+        compdict = BaseSettings(_map_keys(compdict), priority='default')
+        compdict.update(_map_keys(custom))
+    # End backwards support
```
I think we should remove support for the old signature, `build_component_list` is an internal helper for scrapy, there's no need to maintain backward compatibility for it and it makes this function slightly more difficult to understand.

jdemaeyer replied:
I'll have to update a couple of tests but I like dropping this, too

Curita commented on scrapy/utils/conf.py:
```diff
@@ -1,36 +1,54 @@
+            for k, v in six.iteritems(compdict):
+                prio = compdict.getpriority(k)
+                compbs.set(convert(k), v, priority=prio)
+            return compbs
```
It took me a bit to understand this, I think this implementation is halfway between the old behavior and the new one, though I'm probably missing something.

If you call _check_components in a `compdict` that is a BaseSettings instance, you're making sure that no keys convert to the same path/value, regardless of their priority. In that case adding them converted into a new BaseSettings shouldn't make any difference since there aren't any clashes, we can stick to `{convert(k): v for k, v in six.iteritems(compdict)}` (there's no need to return a BaseSettings too).

Another possibility should be to not call _check_components in a BaseSettings, and check manually if there are any two keys that convert to the same path and also have the same priority so we can't tell which should be overridden first (commented in the description of #1267).

It's true that there weren't per-keys priorities before, users working with old paths should get an error if any two keys clash if they switch to scrapy 1.0, but it's probably a nicer deprecation for the old paths to users skipping 1.0 and updating directly to 1.1. If possible I would prefer the second option of checking if any keys that clash have the same priority or not thought I think it's not that critical, we can implement the first one and delete this `if` otherwise.

jdemaeyer replied:
> In that case adding them converted into a new BaseSettings shouldn't make any difference since there aren't any clashes, we can stick to `{convert(k): v for k, v in six.iteritems(compdict)}` (there's no need to return a BaseSettings too).

Keeping track of the priorities (i.e. returning `BaseSettings`) was necessary when there were `base` and `custom` BaseSettings objects with different priorities. But when we drop support for the `(base, custom)` signature you're right, returning a simple dict should be fine.

I'll update this and include the edge case of different paths with different priorities

Curita commented on scrapy/pipelines/__init__.py:
```diff
@@ -21,7 +21,8 @@ def _get_mwlist_from_settings(cls, settings):
             # convert old ITEM_PIPELINE list to a dict with order 500
             item_pipelines = dict(zip(item_pipelines, range(500, 500+len(item_pipelines))))
-        return build_component_list(settings['ITEM_PIPELINES_BASE'], item_pipelines)
+            settings.set('ITEM_PIPELINES', item_pipelines)
```
There are a couple of issues here:
* There's no guarantee that `settings.set('ITEM_PIPELINES', item_pipelines)` overrides the stored ITEM_PIPELINES, that `settings.set()` call should be made with a priority higher than the stored one for that key.
* We didn't modify the settings before, and not too long ago the settings were frozen at this point, this would have thrown an error. A solution for it could be to make a copy and set ITEM_PIPELINES there.
* This is incompatible with the goal of allowing addons to modify settings, addons won't be able to add pipelines easily if ITEM_PIPELINES is a list.

My vote is to remove all this deprecation support for lists, it was introduced way too long ago (0.20 release: fc388f46).

Curita replied:
I recalled that ITEM_PIPELINES can't be a list with the changes in `SettingsAttribute.set()`, ITEM_PIPELINES has as value a BaseSettings instance, if a user tries to set a list here it's going to throw an error when trying to merge them into those BaseSettings.

jdemaeyer replied:
Good point. Providing a list still worked when the `and isinstance(value, Mapping)` was still present in `SettingsAttribute.set()`, but I also prefer dropping support for lists, especially as it will lead to unexpected behaviour for add-on developers.

Curita commented:
I love the unification of the code, really like the design decisions you took there. I pointed out a couple of remaining details but I think the overall functionality is well defined, you should be able to start with the documentation.

jdemaeyer commented:
Alright, I've now incorporated your recent feedback (nitpicks, `BaseSettings` handling in `build_component_list`, remove `custom` support in `build_component_list`, remove non-dict `ITEM_PIPELINES` support), written the documentation and some more tests, rebased into two feature/test/doc-complete commits, and removed the WIP tag. :)

curita commented on scrapy/utils/conf.py:
```diff
@@ -1,36 +1,49 @@
+                                     'object, please update your settings'
+                                     ''.format(compdict))
+                else:
+                    compbs.set(convert(k), v, priority=prio)
```
Simple and clever implementation for checking matching keys after their conversions, really like it.

curita commented on scrapy/utils/conf.py:
```diff
@@ -1,36 +1,49 @@
+                if compbs.getpriority(convert(k)) == prio:
+                    raise ValueError('Some paths in {!r} convert to the same'
+                                     'object, please update your settings'
+                                     ''.format(compdict))
```
Minor detail: Printing `{!r}` for BaseSettings is probably a little bit too verbose, since it prints the SettingsAttributes `repr`s. We could print `list(compdict.keys())` instead (The `list()` call is to convert the `dict_keys` objects in python3 to a proper list).

curita commented on scrapy/settings/__init__.py:
```diff
@@ -88,19 +103,48 @@ def getdict(self, name, default=None):
+            return get_settings_priority('default')
+
+    def __setitem__(self, name, value):
+        self.set(name, value)
```
Really minor detail, but a test for this alias would be nice. Coverage of your changes is 100% otherwise, good work :)

curita commented on tests/test_cmdline/__init__.py:
```diff
@@ -29,3 +30,17 @@ def test_override_settings_using_envvar(self):
         self.assertEqual(self._execute('settings', '--get', 'TEST1'), \
                          'override')
 
+    def test_override_dict_settings(self):
```
Really like to see a test for it :+1:

curita commented:
I think Settings and BaseSetttings docs are great candidates to be used with Sphinx `autodocs`, that way we could automatically document Settings without copying BaseSettings documentation (as we did with CrawlerRunner and CrawlerProcess in #1190), and move forward in bringing docstrings back to source code: #713. Since Settings doesn't implement new methods this is not required right now, so it isn't something to put this pull request on hold for merging by any means.

I saw that you updated the documentation for missing entries in the default settings, seems like we have to take more care when updating them, thanks for that :)

I suggested shamefully minor improvements, but I think the pull request is in great shape and ready to be merged, awesome work Jakob!

jdemaeyer commented:
Pardon the late reply. I have the changes you suggested ready but I guess now that this has a MRG+1 tag they should better go into a new PR?

curita commented:
Let's keep those changes in this pull request, if they implement the changes I suggested my +1 remains.

jdemaeyer commented:
I've changed (and squashed) the [error message content](https://github.com/jdemaeyer/scrapy/blob/enhancement/settings-per-key-priorities/scrapy/utils/conf.py#L29) and added (and squashed) the [``__setitem__`` test](https://github.com/jdemaeyer/scrapy/blob/enhancement/settings-per-key-priorities/tests/test_settings/__init__.py#L94).

There's also a new commit in which I've moved the `Settings`/`BaseSettings` documentation into docstrings. I see that you have used `:inherited-members:` in the `CrawlerProcess` documentation. I have omitted that here because it would be quite a lot of duplication and the `BaseSettings` (+ methods) documentation is directly underneath the small paragraph for `Settings`. Another option might be to use `:inherited-members:` for `Settings` and *don't* use ``:members:`` for `BaseSettings`?

dangra commented:
+1 to merge, It needs a note about backwards incompatibilities introduced by this PR and how to update users code if possible.

jdemaeyer commented:
> +1 to merge, It needs a note about backwards incompatibilities introduced by this PR and how to update users code if possible.

I'm happy to add a note, would that go into `news.rst`? Except for `ITEM_PIPELINES` being no longer allowed to be a list, which was deprecated since `0.20`, there shouldn't be any backwards incompatibilities though. The `BaseSettings` behave just like `dict`s, and the `_BASE` settings still work (and throw a `ScrapyDeprecationWarning` if used).

dangra commented:
Removing item pipelines list support is fine and it is not a backward
incompatible change.

The change that worries me a bit is the new behaviour for dictionary
settings whose values are merged instead or replaced
El ago. 20, 2015 7:05, "Jakob de Maeyer" <notifications@github.com>
escribió:

> +1 to merge, It needs a note about backwards incompatibilities introduced
> by this PR and how to update users code if possible.
>
> I'm happy to add a note, would that go into news.rst? Except for
> ITEM_PIPELINES being no longer allowed to be a list, which was deprecated
> since 0.20, there shouldn't be any backwards incompatibilities though.
> The BaseSettings behave just like dicts, and the _BASE settings still
> work (and throw a ScrapyDeprecationWarning if used).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scrapy/scrapy/pull/1149#issuecomment-132963509>.
>

kmike commented on scrapy/settings/__init__.py:
```diff
@@ -45,64 +64,224 @@ def __str__(self):
+            compsett = BaseSettings(self[name + "_BASE"], priority='default')
+            compsett.update(self[name])
+            return compsett
+        else:
```
else is not necessary, let's drop it

jdemaeyer replied:
Now that I've seen it with a little more distance, I think this whole function doesn't really achieve what I had intended.

When users override `XY_BASE`, they explicitly don't want any of Scrapy's defaults for that component setting. But line 206 pulls Scrapy's defaults (which now live in `XY`) back in, and even worse overwrites the users `XY_BASE` settings where they have the same keys (say when the user simply changed some orders).

I guess either line 206 could be changed so that only those keys from `XY`that have a priority higher than `default` are considered, or support for `_BASE` settings could be dropped altogether with this PR. After all, they have always been marked as "never edit this", there is going to be a release note b/c of the new update-not-overwrite behaviour of the dict-like settings anyways, and we could get rid of this non-public helper function.

kmike commented on scrapy/settings/__init__.py:
```diff
@@ -110,21 +289,96 @@ def setmodule(self, module, priority='project'):
+    def __str__(self):
+        return str(self.attributes)
+
+    __repr__ = __str__
```
I think repr should be something like `Settings(%s) % self.attributes`

kmike replied:
because otherwise it'd be hard to distinguish Settings objects from regular dicts in console

jdemaeyer replied:
Agreed, I'll change that

kmike commented on scrapy/utils/conf.py:
```diff
@@ -1,36 +1,49 @@
-    compdict.update(_map_keys(custom))
-    items = (x for x in six.iteritems(compdict) if x[1] is not None)
-    return [x[0] for x in sorted(items, key=itemgetter(1))]
+def remove_none_values(compdict):
```
This utility function looks generic enough for scrapy.utils.conf - what about moving it to scrapy.utils.python?

Also, I expect `remove_none_values` it to remove values inplace. `without_none_values`? We can also extend it to handle iterables.

jdemaeyer replied:
+1 will change/extend

jdemaeyer commented:
Here's a proposed release note:
```
Dictionary settings are no longer overridden
--------------------------------------------

With this release, Scrapy's dictionary-like settings (e.g. ``ITEM_PIPELINES``,
see full list below) will receive per-key settings priorities (see
:ref:`topics-api-settings`). Internally, this is achieved by promoting them to
:class:`~scrapy.settings.BaseSettings` instances. Instances of this class behave
just like dictionaries, with one notable exception: When being written to, the 
:class:`~scrapy.settings.BaseSettings` instance is *updated*, not overwritten::

    >>> settings.set('ITEM_PIPELINES', {'path.one': 100})
    >>> settings.set('ITEM_PIPELINES', {'path.two': 200})
    >>> print dict(settings['ITEM_PIPELINES'])
    {'path.two': 200, 'path.one': 100}

This facilitates easy enabling, disabling, or configuring of single components.
E.g., if you want to disable the pipeline ``some.pipeline`` from the command
line, it is now sufficient to set the dictionary value for just this pipeline to
``None``::

    scrapy crawl example.com -s ITEM_PIPELINES={"some.pipeline": null}

instead of having to provide the full dictionary of all other enabled pipelines.

This update applies to all dictionary-like settings that have a default value,
namely:

* :setting:`DEFAULT_REQUEST_HEADERS`,
* :setting:`DOWNLOAD_HANDLERS`,
* :setting:`DOWNLOADER_MIDDLEWARES`,
* :setting:`EXTENSIONS`,
* :setting:`FEED_STORAGES`,
* :setting:`FEED_EXPORTERS`,
* :setting:`ITEM_PIPELINES`,
* :setting:`SPIDER_MIDDLEWARES`, and
* :setting:`SPIDER_CONTRACTS`.

If your code relies on completely replacing, not updating, one of these
settings, you can work around the new behaviour by deleting the dictionary-like
setting before writing to it::

    >>> del settings['SPIDER_MIDDLEWARES']
    >>> settings.set('SPIDER_MIDDLEWARES', {'my.middleware': 100})
    >>> print dict(settings['SPIDER_MIDDLEWARES'])
    {'my.middleware': 100}

However, keep in mind that Scrapy's standard procedure to disable components
from the dictionary-like settings is to set their value to ``None``.
```

and another note for the case we decide to drop `_BASE` support (see https://github.com/scrapy/scrapy/pull/1149#discussion_r37554313):
```
Removal of ``_BASE`` settings
-----------------------------

The new update-not-overwrite behaviour of Scrapy's dictionary settings (see
above) renders the ``_BASE`` settings that were previously used to store
Scrapy's defaults (e.g. `DOWNLOADER_MIDDLEWARES_BASE`) obsolete.
Consequentially, Scrapy's defaults have been moved into the regular settings
(e.g. :setting:`DOWNLOADER_MIDDLEWARES`). The ``_BASE`` settings have been
removed, and they are no longer used by Scrapy when building the list of enabled
components.

If your code overrides the ``XY_BASE`` setting, please update it by overriding
(i.e. updating) the ``XY`` setting instead, and setting the value of the default
components that you wish to disable to ``None``.
```

or, if we decide to keep supporting `_BASE` settings despite the new update-not-override behaviour of the dictionary-like settings:
```
Deprecation of ``_BASE`` settings
---------------------------------

The new update-not-overwrite behaviour of Scrapy's dictionary settings (see
above) renders the ``_BASE`` settings that were previously used to store
Scrapy's defaults (e.g. `DOWNLOADER_MIDDLEWARES_BASE`) obsolete.
Consequentially, Scrapy's defaults have been moved into the regular settings
(e.g. :setting:`DOWNLOADER_MIDDLEWARES`).

While setting the ``XY_BASE`` setting will still work, please update your code
to override (i.e. update) the ``XY`` setting instead, and set the value of the
default components that you wish to disable to ``None``.
```

jdemaeyer commented:
Rebased onto current master and updated the function that handles backwards-compatibility for users who explicitly set `_BASE` settings:

They will now find the expected behaviour that they get *none* of Scrapy's defaults for each setting where they manually have set a `_BASE` setting. E.g. doing this:
```
# settings.py
DOWNLOADER_MIDDLEWARES_BASE = {}
```
will now result in no downloader middlewares being enabled (and a warning that `_BASE` settings are deprecated).

Not sure about the codecov test failing. It says only 95 % of this diff are hit because there are a couple of places that weren't covered before but where I switched the syntax to use the new helpers, e.g. like this:
```python
-            valid_output_formats = (
-                list(self.settings.getdict('FEED_EXPORTERS').keys()) +
-                list(self.settings.getdict('FEED_EXPORTERS_BASE').keys())
-            )
+            feed_exporters = without_none_values(self.settings._getcomposite('FEED_EXPORTERS'))
+            valid_output_formats = feed_exporters.keys()
```

These should definitely be covered but I don't think it belongs into this PR.

redapple commented:
Removed "backward-incompatible" tag after https://github.com/scrapy/scrapy/issues/1586 merged
